crates/symbolicator-service/src/metrics.rs

```diff
@@ -1,17 +1,342 @@
+use thread_local::ThreadLocal;
 
-static METRICS_CLIENT: OnceLock<StatsdClient> = OnceLock::new();
+static METRICS_CLIENT: OnceLock<MetricsWrapper> = OnceLock::new();
 
 /// The metrics prelude that is necessary to use the client.
 pub mod prelude {
     pub use cadence::prelude::*;
 }
 
+#[derive(Debug, Clone)]
+struct Sink(Arc<QueuingMetricSink>);
+
+impl MetricSink for Sink {
+    fn emit(&self, metric: &str) -> std::io::Result<usize> {
+        self.0.emit(metric)
+    }
+    fn flush(&self) -> std::io::Result<()> {
+        self.0.flush()
+    }
+}
+
+type LocalAggregators = Arc<ThreadLocal<CachePadded<Mutex<LocalAggregator>>>>;
+
+/// The globally configured Metrics, including a `cadence` client, and a local aggregator.
+pub struct MetricsWrapper {
+    /// The raw `cadence` client.
+    statsd_client: StatsdClient,
+
+    /// A thread local aggregator.
+    local_aggregator: LocalAggregators,
+}
+
+impl fmt::Debug for MetricsWrapper {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("MetricsWrapper")
+            .field("statsd_client", &self.statsd_client)
+            .field(
+                "local_aggregator",
+                &format_args!("LocalAggregators {{ .. }}"),
+            )
+            .finish()
+    }
+}
+
+/// We are not (yet) aggregating distributions, but keeping every value.
+/// To not overwhelm downstream services, we send them in batches instead of all at once.
+const DISTRIBUTION_BATCH_SIZE: usize = 20;
+
+/// Creates [`LocalAggregators`] and starts a thread that will periodically
+/// send aggregated metrics upstream to the `sink`.
+fn make_aggregator(prefix: &str, formatted_global_tags: String, sink: Sink) -> LocalAggregators {
+    let local_aggregators: LocalAggregators = Default::default();
+
+    let aggregators = Arc::clone(&local_aggregators);
+    let prefix = if prefix.is_empty() {
+        String::new()
+    } else {
+        format!("{}.", prefix.trim_end_matches('.'))
+    };
+
+    let thread_fn = move || {
+        let mut formatted_metric = String::with_capacity(256);
+        let mut suffix = String::with_capacity(128);
+        loop {
+            thread::sleep(Duration::from_secs(5));
+
+            let mut total_counters = AggregatedCounters::default();
+            let mut total_distributions = AggregatedDistributions::default();
+
+            for local_aggregator in aggregators.iter() {
+                let (local_counters, local_distributions) = {
+                    let mut local_aggregator = local_aggregator.lock().unwrap();
+                    (
+                        std::mem::take(&mut local_aggregator.aggregated_counters),
+                        std::mem::take(&mut local_aggregator.aggregated_distributions),
+                    )
+                };
+
+                // aggregate all the "counter like" metrics
+                if total_counters.is_empty() {
+                    total_counters = local_counters;
+                } else {
```

Comment:
I don't understand how `total_counters` can ever be nonempty here. We drain it in line 121 and there are no continues in the loop, so it must always be empty at the end of every iteration.

Reply:
I removed the `drain`, as that was not useful.

We have an outer loop running on a timer, and then this inner loop that just aggregates all the thread-local aggregations. This pretty much takes the first thread-local aggregations as a whole instead of going through the merge logic. I think I will move this out into its own function to simplify this a bit.